L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+
+    } else if (creds instanceof ChoiceChannelCredentials) {
+      ChoiceChannelCredentials choiceCreds = (ChoiceChannelCredentials) creds;
+      StringBuilder error = new StringBuilder();
+      for (ChannelCredentials innerCreds : choiceCreds.getCredentialsList()) {
+        Result result = from(innerCreds);
+        if (!result.isError()) {
+          return result;
+        }
+        error.append("; ");
+        error.append(result.getError());
+      }
+      return Result.error(error.substring(2));
+
+    } else {
+      return Result.error(
+          "Unsupported credential type: " + creds.getClass().getName());
+    }
+  }
+
+  public final class Result {
+    public static Result error(String error) {...}
+    public static Result negotiator(ProtocolNegotiator.Factory factory) {...}
+    public Result withCallCredentials(CallCredentials callCreds) {...}
+  }
+}
+```
+
+### ServerCredentials
+
+Server-side will be a clone of client-side, but using the `Server` instead of
+`Channel`. `ServerCredentials` cannot contain `CallCredentials`, so
+`CompositeServerCredentials` will be omitted. There is also currently no known
+use-case for `ChoiceServerCredentials`, so it will be omitted.
+
+```java
+// instead of:
+server = ServerBuilder.forPort(443)
+    .useTransportSecurity(new File("cert.pem"), new File("cert.key"))
+    .build().start();
+server = ServerBuilder.forPort(8080)
+    .build().start();
+// the user would now:
+ChannelCredentials tlsCreds = new TlsServerCredentials(
+    new File("cert.pem"), new File("cert.key"));
+server = Grpc.newServerBuilderForPort(443, tlsCreds)
+    .build().start();
+server = Grpc.newServerBuilderForPort(8080, new InsecureChannelCredentials())
+    .build().start();
+```
+
+
+```java
+package io.grpc;
+
+public abstract class ServerCredentials {}
+```
+
+Since servers using TLS require a certificate, the `TlsServerCredentials`
+cannot have the simple no-arg configuration of `TlsChannelCredentials`. It will
+provide convenience constructors for providing a certificate and unencrypted
+key. If any other configuration is necessary, the `Builder` would be used.
+
+```java
+package io.grpc;
+
+public final class TlsServerCredentials extends ServerCredentials {
+  /**
+   * Creates an instance using provided certificate chain and private key.
+   * Generally they should be PEM-encoded and the key is an unencrypted PKCS#8
+   * key (file headers have "BEGIN CERTIFICATE" and "BEGIN PRIVATE KEY").
+   */
+  public TlsServerCredentials(File certChain, File privateKey)
+      throws IOException {...}
+  // Will not auto-close InputStream, unlike existing transportSecurity()
+  public TlsServerCredentials(InputStream certChain, InputStream privateKey)
+      throws IOException {...}
+
+  /** Same as client-side */
+  public Set<Feature> incomprehensible(EnumSet<Feature> understoodFeatures) {...}
+
+  public byte[] getCertificateChain() { return Arrays.copyOf(...); }
```

Comment:
I have been working on TLS proxying for the past few months and one thing I had to work around a lot was making the Server able to work with SNI.   This means picking the key/cert after most of the other config.   Having just one chain or one key would be limitting.

Reply:
* For advanced use-cases we always want to allow Netty's SslContext to be available
* SNI support is actually something C core has in their SslServerCredential. We could consider supporting it as well in Java. If we did so, then there would be a new Feature that would have a Map or Function for the choosing the certificate.

Reply:
One other minor thing: The SNI builder allows *dynamically* creating a SslContext based on the domain.  There may not be just one.